File: android/activitysources/src/main/java/com/fjuul/sdk/activitysources/entities/internal/healthconnect/HCDataPoint.java
package com.fjuul.sdk.activitysources.entities.internal.healthconnect;

import java.util.Date;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public abstract class HCDataPoint {
    @NonNull
    protected final Date start;
    @Nullable
    protected final Date end;
    @Nullable
    protected final String dataSource;

    @NonNull
    public Date getStart() {
        return start;
    }

    @Nullable
    public Date getEnd() {
        return end;
    }

    @Nullable
    public String getDataSource() {
        return dataSource;
    }

    public HCDataPoint(@NonNull Date start, @Nullable String dataSource) {
        this(start, null, dataSource);
    }

    public HCDataPoint(@NonNull Date start, @Nullable Date end, @Nullable String dataSource) {
        this.start = start;
        this.end = end;
        this.dataSource = dataSource;
    }
} 

File: android/activitysources/src/main/java/com/fjuul/sdk/activitysources/entities/internal/healthconnect/HCDataPoint.kt
package com.fjuul.sdk.activitysources.entities.internal.healthconnect

import java.time.Instant

/**
 * Base class for all Health Connect data points.
 */
abstract class HCDataPoint {
    abstract val start: Instant
    abstract val end: Instant?
    abstract val value: Double
    abstract val dataSources: Set<String>
}

/**
 * Base class for daily data points (e.g. daily steps, daily calories).
 */
abstract class HCDailyDataPoint : HCDataPoint() {
    final override val end: Instant? = null
}

/**
 * Base class for intraday data points (e.g. heart rate samples).
 */
abstract class HCIntradayDataPoint : HCDataPoint() {
    final override val end: Instant
        get() = start
}
